### PR TITLE
HPCC-20895 Refactor Session Management

### PR DESCRIPTION
**Pull request by miguelvazq**
Fixes: HPCC-20895

Signed-off-by: Miguel Vazquez <miguel.vazquez@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [x] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by miguelvazq** [comment]
@GordonSmith Please review.

File: esp/src/src/Session.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import * as xhr from "dojo/request/xhr";
+import * as ESPUtil from "./ESPUtil";
+import * as topic from "dojo/topic"
+
+const espTimeoutSeconds = cookie("ESPSessionTimeoutSeconds") || 600; // 10 minuntes?
+const IDLE_TIMEOUT = espTimeoutSeconds * 10;
+const SESSION_RESET_FREQ = 30 * 1000;
+
+let _prevReset = Date.now();
+let idleWatcher = new ESPUtil.IdleWatcher(IDLE_TIMEOUT);
```
Comment (GordonSmith) [inline esp/src/src/Session.ts]: Should be const?

File: esp/src/src/Session.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+let monitorLockClick = new ESPUtil.MonitorLockClick();
```
Comment (GordonSmith) [inline esp/src/src/Session.ts]: Should be const?

File: esp/src/src/Session.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+let sessionIsActive = espTimeoutSeconds;
```
Comment (GordonSmith) [inline esp/src/src/Session.ts]: Should be const?

File: esp/src/src/Session.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+declare const dojoConfig;
```
Comment (GordonSmith) [inline esp/src/src/Session.ts]: Minor:  by convention I tend to keep the global declares at the top of the file below the imports...

File: esp/src/src/Session.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        });
+
+        idleWatcher.start();
+        monitorLockClick.unlocked();
+    } else if (cookie("ECLWatchUser")) {
+        window.location.replace(dojoConfig.urlInfo.basePath + "/Login.html");
+    }
+}
+
+export function locked() {
```
Comment (GordonSmith) [inline esp/src/src/Session.ts]: rename to "lock" as its an action?

File: esp/src/src/Session.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    idleWatcher.stop();
+}
+
+export function unlocked() {
```
Comment (GordonSmith) [inline esp/src/src/Session.ts]: Rename to "unlock" as its an action?

File: esp/src/src/Session.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    idleWatcher.start();
+}
+
+export function fireIdle() {
+    idleWatcher.fireIdle();
+}
+
+export function _resetESPTime() {
```
Comment (GordonSmith) [inline esp/src/src/Session.ts]: If its exported, then it should probably not have the "_" prefix?  If its only local to this file, then it should not be exported.

**Review by GordonSmith — CHANGES_REQUESTED** [review]
Is this a 7.8 or 7.6 change?

File: esp/src/src/Session.ts (hunk 7 above)
Comment (miguelvazq) [reply esp/src/src/Session.ts]: Good point.

**Comment by miguelvazq** [comment]
@GordonSmith Changes made.

**Review by GordonSmith — APPROVED** [review]
Looks good.

**Comment by miguelvazq** [comment]
> Is this a 7.8 or 7.6 change?

This should go in 7.8

**Pull request merged**